libs/vkd3d/command.c

```diff
@@ -18123,84 +18165,164 @@ static void d3d12_command_queue_execute(struct d3d12_command_queue *command_queu
-     * based on VKD3D_AUTO_CAPTURE_COUNTS.
-     * If a submission index is not marked to be captured after all, we drop any capture here.
-     * Deciding this in the submission thread is more robust than the alternative, since the submission
-     * threads are mostly serialized. */
-    if (debug_capture)
-        debug_capture = vkd3d_renderdoc_command_queue_begin_capture(command_queue);
-#else
-    (void)debug_capture;
-#endif
+            submit = &submit_desc[num_submits++];
+            submit->sType = VK_STRUCTURE_TYPE_SUBMIT_INFO_2;
+            submit->signalSemaphoreInfoCount = 1;
+            submit->pSignalSemaphoreInfos = &binary_semaphore_info;
 
-    if (split_submissions)
-        vr = d3d12_command_queue_submit_split_locked(command_queue->device, vk_queue, num_submits, submit_desc);
-    else if ((vr = VK_CALL(vkQueueSubmit2(vk_queue, num_submits, submit_desc, VK_NULL_HANDLE))) < 0)
-        ERR("Failed to submit queue(s), vr %d.\n", vr);
+            submit = &submit_desc[num_submits++];
+            submit->sType = VK_STRUCTURE_TYPE_SUBMIT_INFO_2;
+            submit->waitSemaphoreInfoCount = 1;
+            submit->pWaitSemaphoreInfos = &binary_semaphore_info;
+        }
 
-    VKD3D_DEVICE_REPORT_FAULT_AND_BREADCRUMB_IF(command_queue->device, vr == VK_ERROR_DEVICE_LOST);
+        if (command_queue->device->vk_info.NV_low_latency2)
+        {
+            spinlock_acquire(&command_queue->device->low_latency_swapchain_spinlock);
+            if ((low_latency_swapchain = command_queue->device->swapchain_info.low_latency_swapchain))
+                dxgi_vk_swap_chain_incref(low_latency_swapchain);
+            consumed_present_id = command_queue->device->frame_markers.consumed_present_id;
+            spinlock_release(&command_queue->device->low_latency_swapchain_spinlock);
+
+            /* If we have submitted a swapchain blit to Vulkan,
+             * it is not possible for a present ID to keep contributing to the frame's completion.
+             * The likely case here is that application just forgot to signal present ID.
+             * Don't bother trying to mark submission present ID if application isn't bothering to set markers properly. */
+            if (low_latency_swapchain && low_latency_frame_id > consumed_present_id &&
+                    dxgi_vk_swap_chain_low_latency_enabled(low_latency_swapchain))
+            {
+                latency_submit_present_info.sType = VK_STRUCTURE_TYPE_LATENCY_SUBMISSION_PRESENT_ID_NV;
+                latency_submit_present_info.pNext = NULL;
+                latency_submit_present_info.presentID = low_latency_frame_id;
+
+                for (i = 0; i < num_submits; i++)
+                    submit_desc[i].pNext = &latency_submit_present_info;
+            }
+
+            if (low_latency_swapchain)
+                dxgi_vk_swap_chain_decref(low_latency_swapchain);
+        }
+
+        if (split_submissions)
```

Comment:
Should not hit this path if submit_count > 1.

Reply:
is this fine even if we have the transition command buffer in there? I'm not sure what the RADV problem we're working around there is exactly.

Reply:
Ah, right ... Well, maybe just leave it as-is. It's there to workaround async compute DGCC issues on RADV where if there's more than one command buffer in the submit bad things happen.